pennylane/qnn.py

```diff
@@ -0,0 +1,205 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the :class:`~.KerasLayer` class for integrating QNodes with the Keras
+layer API."""
+import functools
+import inspect
+from collections.abc import Iterable
+from typing import Optional
+
+try:
+    import tensorflow as tf
+    from tensorflow.keras.layers import Layer
+    from pennylane.interfaces.tf import to_tf
+
+    CORRECT_TF_VERSION = int(tf.__version__.split(".")[0]) > 1
+except ImportError:
+    from abc import ABC
+
+    Layer = ABC
+    CORRECT_TF_VERSION = False
+
+from pennylane.qnodes import QNode
+
+INPUT_ARG = "inputs"
+
+
+class KerasLayer(Layer):
+    """A Keras layer for integrating PennyLane QNodes with the Keras API.
+
+    This class converts a :class:`~.QNode` to a Keras layer. The QNode must have a signature that
+    satisfies the following conditions:
+
+    - Contain an ``inputs`` argument for inputting data. All other arguments are treated as
+      weights within the QNode.
+    - All arguments must accept an array or tensor, e.g., arguments should not use nested lists
+      of different lengths.
+    - All arguments, except ``inputs``, must have no default value.
+    - The ``inputs`` argument is permitted to have a default value provided the gradient with
+      respect to ``inputs`` is not required.
+    - There cannot be a variable number of positional or keyword arguments, e.g., no ``*args`` or
+      ``**kwargs`` present in the signature.
+
+    The QNode weights are initialized within the :class:`~.KerasLayer`. Upon instantiation,
+    a ``weight_shapes`` dictionary must hence be passed which describes the shapes of all
+    weights in the QNode. The optional ``weight_specs`` argument allows for a more fine-grained
+    specification of the QNode weights, such as the method of initialization and any
+    regularization or constraints. If not specified, weights will be added using the Keras
+    default initialization and without any regularization
+    or constraints.
+
+    **Example usage:**
+
+    .. code-block:: python
+
+        n_qubits = 2
+        dev = qml.dev("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev, interface="tf")
+        def circuit(inputs, weights):
+            qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+            qml.templates.StronglyEntanglingLayers(weights, wires=list(range(n_qubits)))
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        weight_shapes = {"weights": (3, n_qubits, 3)}
+        weight_specs = {"weights": {"initializer": "random_uniform"}}
+
+        qnode = qml.qnn.KerasLayer(circuit, weight_shapes, 2, weight_specs)
```

Comment:
Worth having `output_dim=2`, so it's clear from the example what the 2 means?

Reply:
Note: in a couple of docstrings with longer examples, I have experimented with putting the argument list _above_ the example. See https://pennylane.readthedocs.io/en/stable/code/api/pennylane.map.html.

This imitates how NumPy does it: https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.eigvalsh.html

I tend to find this cleaner, as,

* The arguments are listed first, so it makes the example easier to follow
* If you are familiar with the function and just want to double check the signature, you don't have to scroll through the example first.

I'm thinking of expanding that across PL, do you think it makes sense to do it here?

Reply:
I thought we wanted to make UsageDetails the standard (hence my comment above)? Let's take a decision on that soon, before adding more docstrings that need to be fixed.

Reply:
Ah, I've always treated `UsageDetails` sort of like, example overflow? My sweet spot would be:

```rest
function(arg1, arg2)

Brief description of the function.

Args:
    arg1 (type): description
    arg2 (type): description

Returns:
    type: description

**Example:**

Minimal example with 1 or 2 code blocks.

.. UsageDetails::

    More complicated use cases, options, and larger code blocks
```

Reply:
If everyone is in favour, we could make https://github.com/XanaduAI/pennylane/pull/529#discussion_r393037693 the official 'recommended docstring format', and trial it? If successful, make it required.